### PR TITLE
Add Elastic search

### LINKED ISSUE #4875: Integrate Elastic search
**Is your feature request related to a problem? Please describe.**

Integration of Elastic search is important to get closer to a release state. 

- [ ] Check if it is already implemented. If yes, then test it and fix bugs if any.
- [ ] If not implemented, then please implement.

### PR DESCRIPTION
**Pull request by maxlorenz**
Fixes #4875

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:
Adds elasticsearch to the open event server. 

#### Changes proposed in this pull request:

- Adds elasticsearch `docker-compose` entry
- Adds ES config
- Adds first endpoint to search all events

### CODE REVIEW

File: app/views/elastic_cron_helpers.py [hunk 1]
```diff
@@ -19,83 +31,38 @@ def cron_rebuild_events_elasticsearch():
...
-        print(e)
+    deleted, todo = 0, redis_store.scard('event_delete')
+
+    while deleted < todo:
+        deleted += 1
+        event_id = redis_store.spop('event_delete')
+        searchable = SearchableEvent()
+        searchable.meta.id = event_id
+
+        if event_id:  # For safety
```
Comment (schedutron) [inline app/views/elastic_cron_helpers.py]: What is the need of this `if` block?

Comment (maxlorenz) [reply app/views/elastic_cron_helpers.py]: On create, the id is taken from the database and therefore can't be `None`. On delete however, it is possible that redis is out of sync so I double check the id here. It's not completely necessary though

File: app/api/__init__.py [hunk 2]
```diff
@@ -508,3 +509,6 @@
 
 # Admin Sales
 api.route(AdminSalesInvoicesList, 'admin_sales_invoices', '/admin/sales/invoices')
+
+# Search
+api.route(SearchEvent, 'search_event', '/search/event')
```
[the remark this answers is not in the document]
Comment (maxlorenz) [reply app/api/__init__.py]: Okay, I will integrate that into the `/events` endpoint

[the remark this answers is not in the document]
Comment (maxlorenz) [reply app/api/__init__.py]: I put it at `/events/query` now, e.g. `/events/query?name=eventname&location_name=location`

**Comment by niranjan94** [comment]
So, requirements are as follows.

- Accept a filter operation called as `search` as a part of the `filter` query param already being accepted by the `GET /events` endpoint.

	**Example:**
	
	```js
	filter: [
	  {
	    name : 'event',
	    op   : 'search',
	    val  : 'keyword-to-search-for-via-elasticsearch'
	  }
	]
	```
- Use ES to search for it
- Return the event response as usual. (As normal.) The full event response along with any other relationships & filtering/sorting that was requested. 

--

So, basically, this should just supplement what is already there as a part of `/events` endpoint.

**Comment by niranjan94** [comment]
> So basically we would have to search on ES and load the same IDs from Postgres afterwards? I'd prefer to implement full-text search on Postgres before going that route then, what do you think?

Yes. ES will search it's index. And give us the IDs. And we just quickly get the IDs from postgres and return the full response. And, no, we don't want to implement full-text on postgres. ES is meant for these kind of things. Also, we will be using ES to search across sessions too in the future. So, we need ES.

**Comment by maxlorenz** [comment]
I added PG full-text search for events only. Will progress on ES but the solution for PG can be merged and shipped and changed in-place once ES is ready

**Comment by niranjan94** [comment]
> I added PG full-text search for events only. Will progress on ES but the solution for PG can be merged and shipped and changed in-place once ES is ready

@maxlorenz we don't want a PG full-text search. This PR/issue is for ES. And that is exactly what we want to have. So, adding a full-text here doesn't really make sense. So, please add ES instead as per the original requirement.

**Comment by maxlorenz** [comment]
@niranjan94 the search is functional now, at the `/event?search-filter=[{some json}]` as described. `filter` cannot be used since the library would evaluate that to display only events with `search` properties (thus returning nothing every time)

**Comment by mariobehling** [comment]
This is a big topic. Can we proceed with this as WIP and alter things later, please?

@maxlorenz Travis is not building atm.

**Pull request merged**